### PR TITLE
pally: auto-scrub invalid runes from label values

### PR DESCRIPTION
**Pull request by akshayjshah**
In order for vectors to be useful, we need to support auto-scrubbing invalid
runes from label values. For example, we might want to track a count of calls by
service; since service names may contain characters that aren't valid in Tally
(e.g., `!` or 🏦), we should automatically replace those characters with a
fallback.

This is very similar to the logic we use in Muttley.

### CODE REVIEW

File: internal/pally/labels.go [hunk 1]
```diff
@@ -66,3 +69,136 @@ func (d *digester) free() {
...
+
+	return true
+}
+
+// IsValidLabelValue checks whether the supplied string is a valid label value
+// in both Prometheus and Tally.
+//
+// Tally allows label values that match the regexp `^[0-9A-z_\-.]+$`.
+// Prometheus allows any valid UTF-8 string.
+func IsValidLabelValue(s string) bool {
```
Comment (willhug) [inline internal/pally/labels.go]: for simplicity sake, should we just use the same logic for valid names & labels?  I could see people getting confused when they can use '-' in labels, but not metrics.

File: internal/pally/labels.go [hunk 2]
```diff
@@ -66,3 +69,136 @@ func (d *digester) free() {
...
+		}
+	}
+
+	return true
+}
+
+// IsValidLabelValue checks whether the supplied string is a valid label value
+// in both Prometheus and Tally.
+//
+// Tally allows label values that match the regexp `^[0-9A-z_\-.]+$`.
```
Comment (willhug) [inline internal/pally/labels.go]: is there an argument for why we're not using a regex to validate these strings? (easier to read? Efficiency?)

File: internal/pally/labels_test.go [hunk 3]
```diff
@@ -61,3 +64,52 @@ func TestDigester(t *testing.T) {
...
+
+func TestIsValidLabelValue(t *testing.T) {
+	tallyRe := regexp.MustCompile(`^[0-9A-z_.\-]+$`)
+	isValid := func(s string) bool {
+		prom := model.LabelValue(s).IsValid()
+		tally := tallyRe.MatchString(s)
+		return prom && tally
+	}
+
+	assert.NoError(t, quick.CheckEqual(
```
Comment (willhug) [inline internal/pally/labels_test.go]: fancy~
p3ubge2

File: internal/pally/registry.go [hunk 4]
```diff
@@ -70,14 +69,14 @@ func Federated(prom prometheus.Registerer) RegistryOption {
 }
 
 // Labeled adds constant labels to a Registry. All metrics created by a
-// Registry inherit its constant labels.
+// Registry inherit its constant labels. Labels with invalid names or values
+// are dropped.
 func Labeled(ls Labels) RegistryOption {
+	// TODO: Consider whether we should automatically scrub
+	// statically-specified label values.
```
Comment (willhug) [inline internal/pally/registry.go]: +1 for automatically scrubbing, labels simply disappearing will cause a lot more head scratching.

File: internal/pally/labels.go [hunk 5]
```diff
@@ -66,3 +69,136 @@ func (d *digester) free() {
...
+// Prometheus allows any valid UTF-8 string.
```
Comment (willhug) [inline internal/pally/labels.go]: this contradicts the statement above where you say: 
```
Prometheus allows the regexp `^[A-z_:][0-9A-z_:]*$` for metric names, and `^[A-z_][0-9A-z_]*$` for label names.
```

Should we just use `^[A-z_][0-9A-z_]*$` for everything?

File: internal/pally/labels.go (hunk 2 above)
Comment (akshayjshah) [inline internal/pally/labels.go]: Efficiency. This is much, much faster, and it's the performance-gating operation.

File: internal/pally/labels.go (hunk 5 above)
Comment (akshayjshah) [inline internal/pally/labels.go]: I don't think it does. Prom metrics are, conceptually, 

```
type LabelName string
type LabelValue string

type metadata struct {
  name string
  labels map[LabelName]LabelValue
}
```

The rules for label names and label values are different.

File: internal/pally/labels.go (hunk 1 above)
Comment (akshayjshah) [inline internal/pally/labels.go]: See above.

File: internal/pally/registry.go (hunk 4 above)
Comment (akshayjshah) [inline internal/pally/registry.go]: This is actually more generous than M3's behavior - it just silently drops any metrics with invalid characters. I'm happy to revisit this later, but would rather not do so now.

File: internal/pally/labels.go (hunk 5 above)
Comment (willhug) [inline internal/pally/labels.go]: ooooooooooh, I see now, my bad.

**Pull request merged**